File: Algorithm/Code_Materials/c_or_cpp/charpter17_greedy/level2/merge.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.size() == 0) {
            return {};
        }
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> merged;
        for (int i = 0; i < intervals.size(); ++i) {
            int L = intervals[i][0], R = intervals[i][1];
            if (!merged.size() || merged.back()[1] < L) {
                merged.push_back({L, R});
            }
            else {
                merged.back()[1] = max(merged.back()[1], R);
            }
        }
        return merged;
    }
 

int main() {
    int[][] intervals = {{1, 3}, {2, 5}, {4, 6}};
    int[][] merged = merge(intervals);
    for (int i = 0; i < merged.length; i++) {
        for (int j = 0; j < merged[i].length; j++) {
            cout << merged[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}

File: Algorithm/Code_Materials/python_source/charpter17_greedy/level2/merge.py
# coding=utf-8

class Merge:
    def merge(self, intervals):
        intervals.sort(key=lambda x: x[0])
        merged = []
        for interval in intervals:
            # 如果列表为空，或者当前区间与上一区间不重合，直接添加
            if not merged or merged[-1][1] < interval[0]:
                merged.append(interval)
            else:
                # 否则的话，我们就可以与上一区间进行合并
                merged[-1][1] = max(merged[-1][1], interval[1])
        return merged


if __name__ == '__main__':
    intervals = [[1, 3], [2, 6], [8, 10], [15, 18]]
    merge = Merge()
    print merge.merge(intervals)
